Add new IndieAuth server functions

capjamesg opened the pull request:
This PR contains my work porting over my IndieAuth spec implementations to the indieweb-utils library. In the commits in this PR, I have:

1. Added helper functions to validate an authorization and token request.
2. Added functions to process an authorization and token request (i.e. generate / decode a token).
3. Added documentation for a few of the functions listed.

I have not yet added full test coverage for the additions in this commit. I also need to add documentation for a few functions. But, the code in this PR is a good start. I'm open to accepting contributions on the indieauth_server_functions branch for documenting the functions in the indieauth directory.

I have created this change as a PR at this stage as there is a lot of code to review and this change will likely take some time for me to work on.

capjamesg committed:
begin writing indieauth server functions

capjamesg committed:
work on context changes

jamesvandyne commented on src/indieweb_utils/indieauth/profile.py:
```diff
@@ -0,0 +1,80 @@
+    except:
+        raise ProfileError("Request to retrieve profile URL did not return a valid response.")
+
+    if not me_profile:
+        return Profile(
+            name="",
+            photo="",
+            url="",
+            email=""
+        )
```
In what case would me_profile be `None`? I'd expect it would always exist because errors are raising a ProfileError above. Can probably delete this?

jamesvandyne commented on src/indieweb_utils/indieauth/server.py:
```diff
@@ -0,0 +1,308 @@
+        :rtype: str
+    """
+
+    if not client_id or not redirect_uri or not response_type or (not state and state != ""):
```
According to the typehints and docstrings we'll always be passing in strings for every parameter, so we could simplify this.

```suggestion
    if all([client_id, redirect_uri,  response_type, state]):
```

jamesvandyne commented on src/indieweb_utils/indieauth/server.py:
```diff
@@ -0,0 +1,308 @@
+    if not client_id or not redirect_uri or not response_type or (not state and state != ""):
+        raise AuthenticationError("Token request is missing required parameters.")
+
+    if response_type != "code" and response_type != "id":
```
```suggestion
    if response_type not in ["code", "id"]:
```

jamesvandyne commented on src/indieweb_utils/indieauth/server.py:
```diff
@@ -0,0 +1,308 @@
+    if response_type != "code" and response_type != "id":
+        raise AuthenticationError("Only code and id response types are supported.")
+
+    random_string = "".join(random.choice(string.ascii_uppercase + string.digits) for _ in range(10))
```
What is random_string used for? I don't recall reading about that in the spec?  I'd suggest we make a function to generate random strings / tokens for us and use that throughout.

```python
import binascii
import os

def generate_token(*, size: int = 20) -> str:
    return binascii.hexlify(os.urandom(size)).decode()
```

This would return a value like `"48a0ea3a0df4bc46c2bc13f8d623c963c9009919"` (generally the returned value is twice passed the size)

capjamesg replied:
Good question. That code may be redundant. I added that section per:

> The client first creates a code verifier for each authorization request by generating a random string using the characters [A-Z] / [a-z] / [0-9] / - / . / _ / ~ with a minimum length of 43 characters and maximum length of 128 characters. This value is stored on the client and will be used in the authorization code exchange step later.

This section is described fully in the "[Authorization Request](https://indieauth.spec.indieweb.org/#authorization-request)" part of the spec.

But, it looks like this value is not separate from the code verifier which is already generated separately.

capjamesg replied:
After further review, this random string should be code_verifier. In my server-side implementation, I save code_verifier in a session object for later use. I then compare the decoded challenge with the code_verifier during the authorization stage to confirm the integrity of the request. Although that does mean that we should expose the random string in the generate_auth_token function. I shall include that in my next commit.

capjamesg committed:
response to feedback in pr 28

capjamesg committed:
add new documentation for authentication functions

capjamesg committed:
add new documentation for authentication functions

capjamesg committed:
fix autofunction rendering

jalcine commented on docs/source/indieauth.rst:
```diff
@@ -0,0 +1,133 @@
+4. Verifies the length of the code challenge is within the range of 43 and 128 characters.
+
+You should use this function to ensure a POST request to an authorization endpoint to
+redeem an authorization code (per 5.3 Redeeming the Authorization Code in the IndieAuth spec)
```
Can this be made into a link to the specific part?

https://indieauth.spec.indieweb.org/#redeeming-the-authorization-code

jalcine commented on src/indieweb_utils/indieauth/server.py:
```diff
@@ -0,0 +1,325 @@
+        :rtype: str
+    """
+
+    if all([client_id, redirect_uri, response_type, state]):
```
Wouldn't this be `if not all(..)` in the event that one of these values are missing?

jalcine reviewed (commented):
From a cursory glance, this looks good!

capjamesg committed:
add link in docs, fix all() bug

capjamesg commented:
@jalcine Good catch re: the if not all(..) issue. I have addressed that in my latest commit. I have also added a link to the section you mentioned. Thank you for giving the PR a look!

Commit:
Add constants for css selector names

Commit:
Refactor profile extraction

This commit refactors the extraction logic to make it easier to modify
extraction for a single parameter without effecting the other parameters.

Commit:
Add jwt dependency

Commit:
Merge pull request #29 from capjamesg/indieauth_refactor_profile

Indieauth refactor profile

capjamesg commented:
@jamesvandyne Thank you for your PR to refactor the Profile method and add new tests. I noticed a function called "test_handles_missingadasd()" in the tests file. What is the purpose of that function? The function looks to be the same as "test_handles_missing_photo()" with a different name.

capjamesg committed:
add test coverage for indieauth functions

capjamesg committed:
finish test coverage for indieauth utils

capjamesg commented:
@jamesvandyne I have just added tests to accomodate for all of the new IndieAuth utility functions in the library. I had to make an adjustment to the code challenge logic as it was incorrect. All of the test cases are running on my machine with no issues. Please review.

capjamesg committed:
fix linting issues, fix mismatched data types

capjamesg committed:
add reply context test case

capjamesg commented:
Commit 5dd7bb0 introduces a new test case for the reply context feature. I now realise this is out of scope for this PR. We can move the code to a new branch for further review or include it here since the test case is small and passes. The test case does rely on a HTTP network request, however. This behaviour can and should be replaced with the HTML we have in the article.html file.

jamesvandyne commented on src/indieweb_utils/indieauth/server.py:
```diff
@@ -0,0 +1,333 @@
+    code_challenge: str,
+    code_challenge_method: str,
+    allowed_methods: list = ["S256"]
+) -> bool:
```
As this will raise an exception with any error i.e. never returns `False`, do we need to return True or anything at all? If we want to return `True`, maybe we should rename the function to better match the return type? `is_authorization_response_valid` or so?

jamesvandyne commented on src/indieweb_utils/indieauth/server.py:
```diff
@@ -0,0 +1,333 @@
+        if code_challenge != decoded_code["code_challenge"]:
+            raise AuthenticationError("Code challenge in decoded code was invalid.")
+
+    message = _verify_decoded_code(
```
```suggestion
    valid = _verify_decoded_code(
```

jamesvandyne commented on src/indieweb_utils/indieauth/server.py:
```diff
@@ -0,0 +1,333 @@
+        decoded_code["expires"]
+    )
+
+    if message is False:
```
```suggestion
    if not valid:
```

jamesvandyne commented on tests/test_indieauth/test_profile.py:
```diff
@@ -0,0 +1,130 @@
+
+
+@responses.activate
+def test_handles_missingadasd(name_tag):
+    url = "http://example.com"
+    responses.add(
+        responses.Response(
+            method="GET",
+            url=url,
+            body='<div class="h-card"><a class="u-email"></a></div>',
+        )
+    )
+    actual = profile.get_profile(me=url)
+    assert actual.photo is None
```
I accidentlly committed this when I was testing how it would behave if we had a malformed a tag (i.e. no href . Please delete :)

jamesvandyne commented on tests/test_indieauth/test_verification.py:
```diff
@@ -0,0 +1,57 @@
+
+class TestGenerateAuthToken:
+    def test_generate_auth_token(self):
+        """Test function to generate an authentication token."""
```
Might be better to add some more explanation of the test in the doc strings.
```suggestion
        """A end-to-end happy path test to confirm the entire indieauth flow is working properly.
         It will: generate an auth token, redeem the token, validate access, and validate authorization.
        """
```

jamesvandyne reviewed (approved):
LGTM! Nice work!

capjamesg committed:
Add new IndieAuth test description

Co-authored-by: James Van Dyne <james@sugoisoft.com>

capjamesg committed:
respond to feedback on pr

capjamesg committed:
Merge branch 'main' into indieauth_server_functions

capjamesg committed:
change test case to look for noen
